syft/rekor/query.go

```diff
@@ -0,0 +1,182 @@
+package rekor
+
+import (
+	"context"
+	"crypto"
+	"crypto/sha1"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/source"
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/client/entries"
+	"github.com/sigstore/rekor/pkg/generated/client/index"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/spdx/tools-golang/spdx"
+)
+
+// getSbom attempts to retrieve the SBOM from the URI.
+//
+// Precondition: att is not nil and client is not nil
+func getSbom(att *InTotoAttestation, client *http.Client) (*[]byte, error) {
+	if len(att.Predicate.Sboms) > 1 {
+		log.Info("Attestation found on Rekor with multiple SBOMS, which is not currently supported. Proceeding with the first SBOM.")
+	}
+	uri := att.Predicate.Sboms[0].Uri
+
+	resp, err := client.Get(uri)
+	if err != nil {
+		return nil, fmt.Errorf("error making http request: %w", err)
+	}
+
+	if resp == nil {
+		return nil, errors.New("http response body is nil")
+	}
+	bytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading http response: %w", err)
+	}
+
+	return &bytes, nil
+}
+
+// getUuids returns the uuids of the Rekor entries associated with an sha hash
+//
+// Precondition: client is not nil
+func getUuids(sha string, client *client.Rekor) ([]string, error) {
+	query := &models.SearchIndex{Hash: sha}
+	var params *index.SearchIndexParams = index.NewSearchIndexParams().WithQuery(query)
+
+	res, err := client.Index.SearchIndex(params)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, errors.New("result of Rekor search is nil")
+	}
+	return res.Payload, nil
+}
+
+// Precondition: Client is not nil
+func getRekorEntry(uuid string, client *client.Rekor) (models.LogEntry, error) {
+	params := entries.NewGetLogEntryByUUIDParams().WithEntryUUID(uuid)
+	res, err := client.Entries.GetLogEntryByUUID(params)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, errors.New("result of Rekor query is nil")
+	}
+	return res.Payload, nil
+}
+
+// Precondition: client and its fields are not nil
+func getAndVerifySbomFromUuid(uuid string, client *Client) (*sbomWithDigest, error) {
+	logEntry, err := getRekorEntry(uuid, client.rekorClient)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving Rekor entry by uuid %v: \n\t\t%w", uuid, err)
+	}
+	if len(logEntry) == 0 {
+		return nil, fmt.Errorf("retrieved Rekor entry has no logEntryAnons")
+	}
+
+	// logEntry is a map from uuids to logEntryAnons
+	var logEntryAnon models.LogEntryAnon
```

Comment:
if you are expecting only a single rekor entry (which is the case, there should only be 1 per uuid, and this for loop suggests that too), then  above you may want to check that `len(logEntry) == 1`.